[TOOLS-4726] add comment migration function in all source DB types

Kang-dot opened the pull request:
<http://jira.cubrid.org/browse/TOOLS-4726>

### Purpose
CMT의 모든 source db에 대해 table, table column, view, view column에 대해 comment를 migration할 수 있도록 수정합니다.

### Implementation
Oracle
- 기존에 comment이관 기능이 지원되었기 때문에 추가 수정 없음

MySQL
- table, table comment 이관 기능 추가
- mysql은 view, view column comment 지원하지 않음(column의 경우 원본 테이블 컬럼의 comment를 따라감)

MSSQL
- table, table comment, view, view comment 이관 기능 추가

CUBRID
- table comment 이관 기능 추가

informix
- comment 기능 미지원

### Remarks
N/A

Kang-dot committed:
[TOOLS-4726] aply comment migration in all source db types

Kang-dot committed:
[TOOLS-4726] add comment migration in cubrid

- add table, table column, view, view column comment migration function

Kang-dot committed:
[TOOLS-4726] modify abstract schema fetcher

Kang-dot committed:
[TOOLS-4726] remove setComment from abstract Schema fetcher

- apply formatter

Kang-dot committed:
[TOOLS-4726] add method description

Kang-dot committed:
[TOOLS-4726] modify codes

- add log code line
- change query to static string

Kang-dot committed:
Merge branch 'develop' into TOOLS-4726_2503

Kang-dot committed:
[TOOLS-4726] add view column migration function in mssql

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/core/dbmetadata/AbstractJDBCSchemaFetcher.java:
```diff
@@ -1294,4 +1294,27 @@ protected String commentEditor(String comment) {
+     * @return String table comment
+     * @throws SQLException e
+     */
+    protected abstract String getTableComment(Connection conn, String SchemaName, String tableName)
```
```suggestion
    protected abstract String getTableComment(Connection conn, String schemaName, String tableName)
```

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/core/dbmetadata/AbstractJDBCSchemaFetcher.java:
```diff
@@ -1294,4 +1294,27 @@ protected String commentEditor(String comment) {
+     * @return String view comment
+     * @throws SQLException e
+     */
+    protected abstract String getViewComment(Connection conn, String SchemaName, String viewName)
```
```suggestion
    protected abstract String getViewComment(Connection conn, String schemaName, String viewName)
```

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mariadb/meta/MariaDBSchemaFetcher.java:
```diff
@@ -226,6 +232,7 @@ public Catalog buildCatalog(final Connection conn, ConnParameters cp, IBuildSche
 
             for (Table table : tableList) {
                 table.setDDL(getTableDDL(conn, table.getName()));
+                table.setComment(getTableComment(conn, catalog.getName(), table.getName()));
```
table에서는 `catalog.getName()`, view에서는 `schema.getName()`(244 라인) 을 사용하고 있는데 차이가 있는 것인가요?

Kang-dot replied:
삭제해야 하는 코드 입니다. mariaDB의 경우 view comment를 지원하지 않기 때문에 해당 라인은 필요가 없습니다. 삭제하여 내용 반영하겠습니다

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/cubrid/meta/CUBRIDSchemaFetcher.java:
```diff
@@ -2506,4 +2509,18 @@ private boolean getPrivilege(Connection conn, Catalog catalog) {
     }
+
+    @Override
+    protected String getTableComment(Connection conn, String SchemaName, String tableName)
```
```suggestion
    protected String getTableComment(Connection conn, String schemaName, String tableName)
```

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/cubrid/meta/CUBRIDSchemaFetcher.java:
```diff
@@ -2506,4 +2509,18 @@ private boolean getPrivilege(Connection conn, Catalog catalog) {
+    }
+
+    @Override
+    protected String getViewComment(Connection conn, String SchemaName, String viewName)
```
```suggestion
    protected String getViewComment(Connection conn, String schemaName, String viewName)
```

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mariadb/meta/MariaDBSchemaFetcher.java:
```diff
@@ -488,24 +496,26 @@ protected void buildTableColumns(
         if (version.getDbMajorVersion() >= 5) {
             sqlStr =
-                    "SELECT COLUMN_NAME, CHARACTER_SET_NAME "
+                    "SELECT COLUMN_NAME, CHARACTER_SET_NAME, COLUMN_COMMENT "
                             + "FROM INFORMATION_SCHEMA.COLUMNS "
                             + "WHERE TABLE_SCHEMA=? AND TABLE_NAME=?";
 
             try {
                 stmt = conn.prepareStatement(sqlStr);
-                stmt.setString(1, schema.getName());
+                stmt.setString(1, catalog.getName());
```
변경의 이유가 있나요? 이전 코드가 잘못 동작하고 있던 것인가요?

Kang-dot replied:
mariaDB는 oracle과 스키마 구조가 다릅니다 oracle은 db안에 schema(사용자)가 있고 그 하위에 오브젝트가 있다면 mariaDB는 schema = DB이며 DB안에 오브젝트가 있고 사용자는 그 오브젝트를 조회할 권한을 가질 뿐입니다. 이 때문에 mariaDB로 이관을 하면 schema는 단순 계정 이름이 되고, 실제 DB는 catalog가 됩니다. mysql도 같은 맥락으로 동일한 수정 사항을 적용했습니다

<img width="850" height="310" alt="image" src="https://github.com/user-attachments/assets/5f59a971-0df9-4515-b1eb-f2a548565931" />


schema 이름, 즉 유저 접속 계정명으로 조회하면 당연히 아무것도 조회가 되지 않습니다 아마 오작동하고 있던것으로 추측됩니다.

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mariadb/meta/MariaDBSchemaFetcher.java:
```diff
@@ -990,6 +1000,58 @@ protected String getTableDDL(final Connection conn, final String tableName)
+     * @throws SQLException e
+     */
+    @Override
+    protected String getViewComment(Connection conn, String SchemaName, String viewName)
```
```suggestion
    protected String getViewComment(Connection conn, String schemaName, String viewName)
```

Srltas replied:
수정이 안 됐습니다.

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mysql/meta/MySQLSchemaFetcher.java:
```diff
@@ -969,6 +978,58 @@ protected String getTableDDL(final Connection conn, final String tableName)
+     * @throws SQLException e
+     */
+    @Override
+    protected String getViewComment(Connection conn, String SchemaName, String viewName)
```
```suggestion
    protected String getViewComment(Connection conn, String schemaName, String viewName)
```

Srltas replied:
수정이 안 됐습니다.

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mariadb/meta/MariaDBSchemaFetcher.java:
```diff
@@ -990,6 +1000,58 @@ protected String getTableDDL(final Connection conn, final String tableName)
+            throw new IllegalArgumentException("The table name is null!");
+        }
+
+        PreparedStatement stmt = null; // NOPMD
+        ResultSet rs = null; // NOPMD
+        try {
+            LOG.debug("[SQL]{} (1={}, 2={})", SQL_GET_TABLE_COMMENT, schemaName, tableName);
+            stmt = conn.prepareStatement(SQL_GET_TABLE_COMMENT);
+            stmt.setString(1, schemaName);
+            stmt.setString(2, tableName);
+            rs = stmt.executeQuery();
+
+            String comment = null;
+
+            while (rs.next()) {
+                comment = rs.getString(2);
+            }
+
+            return comment;
+        } finally {
+            Closer.close(rs);
+            Closer.close(stmt);
+        }
```
NIT: [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)를 사용해 보는 것은 어떤가요?

Kang-dot replied:
확인했습니다 한번 수정해보겠습니다

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mssql/meta/MSSQLSchemaFetcher.java:
```diff
@@ -115,6 +115,79 @@ public final class MSSQLSchemaFetcher extends AbstractJDBCSchemaFetcher {
+                    + "        'ms_user_view1',  "
+                    + "        'COLUMN',       "
+                    + "        'varchar_col'         "
+                    + "    )";
```
바인드 변수를 넣어야 하는데 `mssq_user`, `ms_user_view1`, `varchar_col`이 들어가 있습니다.

Kang-dot replied:
확인했습니다 수정하겠습니다

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mssql/meta/MSSQLSchemaFetcher.java:
```diff
@@ -441,6 +519,9 @@ protected void buildTableColumns(
                 column.setDefaultValue(defaultValue);
                 String shownDataType = dtHelper.getShownDataType(column);
                 column.setShownDataType(shownDataType);
+                column.setComment(
+                        getTableColumnComment(
+                                conn, schema.getName(), table.getName(), column.getName()));
```
`getTableColumnComment()` 내부에서 컬럼 하나당 쿼리를 한 번씩 실행하여 주석을 조회하고 있어, N+1 쿼리 문제가 발생합니다. 이전에도 이와 같은 문제가 있어 수정한 적이 있습니다. (#126)

MSSQL의 경우 JDBC로 컬럼을 조회하고 있으니 테이블 컬럼의 주석만 한번에 조회하고 컬럼 생성 루프에서 컬럼마다 주석을 매핑해 주는 방식은 어떤가요?

Kang-dot replied:
jdbc로 컬럼을 조회한다는게 getMetadata를 활용하고 있다는 말인지 정확히 이해를 하지 못했습니다. 다만 mssql은 comment기능이 table, view에 있는것이 아니라 extended properties라는 다른 기능에 정의되어 있어서 metadata에서 가져올 수 없습니다. 다만 개선 방안에 대해서는 좀 더 고민해보겠습니다

Srltas replied:
네 컬럼 정보를 얻기 위해 getMetadata를 활용하고 있다는 말이었습니다. 조금 헷갈리게 작성했네요.
getMetadata를 통해 comment를 조회하는 것은 한계가 있으니, 쿼리로 직접 테이블 모든 컬럼의 주석을 **한번만 조회**하자는 얘기입니다.
현재 동작을 보면 컬럼 하나하나 쿼리를 날려 하나씩 comment를 조회하고 있는 것은 성능에 문제가 될 것 같습니다.

Kang-dot committed:
edit review feature

- apply low camel case
- use try-with-resource
- edit bind parameter
- edit mssql table column comment method

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mssql/meta/MSSQLSchemaFetcher.java:
```diff
@@ -484,6 +594,8 @@ protected void buildViewColumns(
         for (Column column : view.getColumns()) {
             String shownDataType = dtHelper.getShownDataType(column);
             column.setShownDataType(shownDataType);
+            column.setComment(
+                    getViewColumnComment(conn, schema.getName(), view.getName(), column.getName()));
```
View 컬럼의 주석 또한 테이블과 같은 방식으로 한번에 조회할 수 있지 않을까요?

Kang-dot replied:
확인했습니다. 수정하겠습니다

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mssql/meta/MSSQLSchemaFetcher.java:
```diff
@@ -115,6 +115,101 @@ public final class MSSQLSchemaFetcher extends AbstractJDBCSchemaFetcher {
+            "SELECT "
+                    + "    SCHEMA_NAME(v.schema_id)   AS schema_name, "
+                    + "    v.name                     AS view_name, "
+                    + "    ep.value                   AS view_comment "
```
해당 SQL을 사용하는 메소드를 보면 조회된 정보에서  `view_comment`만 사용하고 있는데 `schema_name`, `view_name`을 같이 조회할 필요는 없을 것 같습니다.

Kang-dot replied:
네 확인했습니다. 수정하겠습니다

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mssql/meta/MSSQLSchemaFetcher.java:
```diff
@@ -885,4 +997,200 @@ protected void buildTableFKs(
+        } finally {
+            Closer.close(rs);
+            Closer.close(stmt);
+        }
```
`MSSQLSchemaFetcher`에 있는 코드들도 try-with-resources를 적용하는 것이 좋아 보입니다.

Kang-dot replied:
확인했습니다. 수정사항을 적용합니다.

Kang-dot committed:
edit mssql schema fetcher file

- remove non-used column in query
- add try-with-catch

Kang-dot committed:
apply formatter

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mssql/meta/MSSQLSchemaFetcher.java:
```diff
@@ -885,4 +1008,233 @@ protected void buildTableFKs(
+     * @param tableName Table name
+     * @param columnName Column name
+     */
+    protected String getTableColumnComment(
+            Connection conn, String schemaName, String tableName, String columnName)
```
해당 메서드는 더 이상 사용되지 않는 것으로 보입니다.
특별한 이유가 없다면 제거해도 괜찮을 것 같습니다.

Srltas commented on plugins/com.cubrid.cubridmigration.core/src/com/cubrid/cubridmigration/mssql/meta/MSSQLSchemaFetcher.java:
```diff
@@ -885,4 +1008,233 @@ protected void buildTableFKs(
+    }
+
+    protected String getViewColumnComment(
+            Connection conn, String schemaName, String viewName, String viewColumnName)
```
`getTableColumnComment`와 동일해 보입니다.

Kang-dot committed:
remove unused method
